File: packages/generator/src/processing/skimage_bridge.py
"""Bridge script for scikit-image skeletonization methods.

Called from TypeScript via `uv run --with scikit-image`.
Reads JSON from stdin: {width, height, method, bitmap (base64), maxIter?}
Writes JSON to stdout: {skeleton (base64)}
"""

import base64
import json
import sys

import numpy as np
from skimage.morphology import medial_axis, skeletonize, thin

data = json.loads(sys.stdin.read())
width = data["width"]
height = data["height"]
method = data["method"]
raw = base64.b64decode(data["bitmap"])
image = np.frombuffer(raw, dtype=np.uint8).reshape(height, width) > 0

if method == "lee":
    result = skeletonize(image, method="lee")
elif method == "zhang":
    result = skeletonize(image, method="zhang")
elif method == "thin":
    max_iter = data.get("maxIter")
    result = thin(image, max_num_iter=max_iter)
elif method == "medial-axis":
    result = medial_axis(image)
else:
    result = skeletonize(image)

output = result.astype(np.uint8)
print(json.dumps({"skeleton": base64.b64encode(output.tobytes()).decode()}))

File: packages/generator/src/processing/skimage-bridge.ts
import 'bun';
import { resolve } from 'node:path';

const SCRIPT_PATH = resolve(import.meta.dirname, 'skimage_bridge.py');

export type SkimageMethod = 'lee' | 'thin' | 'zhang' | 'medial-axis';

/**
 * Call scikit-image skeletonization via a Python subprocess.
 * Requires `uv` (https://docs.astral.sh/uv/) to be installed.
 * scikit-image is installed automatically on first run via `uv run --with`.
 */
export async function skimageSkeletonize(
  bitmap: Uint8Array,
  width: number,
  height: number,
  method: SkimageMethod,
  maxIter?: number,
): Promise<Uint8Array> {
  const input = JSON.stringify({
    width,
    height,
    method,
    bitmap: Buffer.from(bitmap).toString('base64'),
    ...(maxIter != null && { maxIter }),
  });

  const proc = Bun.spawn(['uv', 'run', '--with', 'scikit-image', '--with', 'numpy', 'python3', SCRIPT_PATH], {
    stdin: new Blob([input]),
    stdout: 'pipe',
    stderr: 'pipe',
  });

  const [stdout, stderr, exitCode] = await Promise.all([new Response(proc.stdout).text(), new Response(proc.stderr).text(), proc.exited]);

  if (exitCode !== 0) {
    throw new Error(`scikit-image bridge failed (exit ${exitCode}): ${stderr.trim()}`);
  }

  const result = JSON.parse(stdout);
  const decoded = Buffer.from(result.skeleton, 'base64');
  return new Uint8Array(decoded);
}
